fecgov/fec-cms
# feature/3688 admin fine calc

## Summary

- Resolves #3688 
- Resolves #3673 
- Builds on 3673, which is still waiting for reviews so I'm moving everything there to here
- Adds the admin fines calculator option to the base template of the site
- Adds admin fine calculations (3688 was only the visual parts)
- Adds automated tests for said calculations

## Impacted areas of the application
- Added code to `base.html` to detect the calcaf class name and include its modal code if needed
- `calc-admin-fines.modal.js` runs this modal, based on other modals on the site
- `calc-admin-fines.js` runs the calculator itself
- `_calc-admin-fines.scss` are the styles specifically for the calculator
- Added ☝️ to `base.scss` so it's part of the base template (We may want to separate those before launch. Still undecided.)
- Added `calc-admin-fines.js` to the Webpack build
- Added `calc-admin-fines-logic.js` to separate the logic (may restructure that file so it's not a component but wanted to get reviews started)
- Added `tests/js/calc-admin-fines.js` for unit (code) tests (part of `npm run test-single`)

## Screenshots

![image](https://user-images.githubusercontent.com/26720877/80217043-0a13fb80-860d-11ea-8893-6e42e33e6656.png)

Successful `npm run test-single`:
![image](https://user-images.githubusercontent.com/26720877/80593561-72892100-89ef-11ea-9f70-b6a24f85fa6b.png)

## Related PRs

branch | PR
------ | ------
feature/3673-admin-fines-calc | [#3712 ]() (closing this)

## How to test
- Pull the branch
- `npm i`
- `npm run build`
- **Add the Vue CDN to the approved CORS domains**
  - Edit `fec/fec/middleware.py`, line 14 (the `script-src` entry)
  - Update `digitalgov.gov",` to `digitalgov.gov https://cdn.jsdelivr.net",`
  - Save your edit (after testing, feel free to discard this change)
  - No reason to commit this yet. That CDN and approach still TBD. Thoughts?
- `./manage.py runserver`
- Add a link to a page
  - In your 127.0.0.1 / localhost, go into [Wagtail](http://127.0.0.1:8000/admin) and edit a page—any page that uses the base.html template. I know that's hard to quickly find/decipher. I used `http://127.0.0.1:8000/admin/pages/7600/edit/`
  - Add some html somewhere: `<a href="" class="js-admin-fines-calc-modal">Launch the Administrative Fine Calculator now</a>` I added a new html block to the page and put my code in there. You can handle it however you like as long as the class is the same (may need to be an anchor tag but maybe not)
  - Save Draft, [Preview](http://127.0.0.1:8000/admin/pages/7600/edit/preview/)
  - Your edit / addition should be on the page
- Click your link and the calculator should open in a modal
- Try to break it!
- Is it clear?
- Is it easy? Fun?
- **Time to test the math!**
- [Open a calculator or three](https://www.fec.gov/legal-resources/enforcement/administrative-fines/calculating-administrative-fines/#calculating-a-fine)
- Enter the same values/scenarios into the old calculator and compare with what you get from the new one (the three old/current calculators correspond to 2017, 2018, and 2019 in the new one)
- Important testing logic (for testing the details) :
  - I added 2020 but its values haven't been approved yet so the math for 'I haven't been assessed a fine', 'TBD 2020', and 2019 should all turn out the same. Technically 'I haven't been assessed a fine' will use the most recent year. The years come from the logic file's `availableDates`
  - Receipts and disbursements are added together to get the fine level ($1 receipts with $1 disbursements = $2 total ≠ $0)
  - A $0 total r & d = $0 fine, even if 1000 days late
  - Check the logic file's `values` for dollar steps (the maxRD values)
    - No reason to test the same group more than twice
    - If you'd like, test $1 and $2 but everything < $5000 is the same test (as far as financial activity goes)
    - Same with $1,000,000+
- Check various combinations of
  - years
  - election-sensitivity
  - late / non-filed
  - days late (if late and not not-filed)
  - previous violations
  - [receipts + disbursements] values
- Run `npm run test-single` to make sure the tests pass (if you're able to build)
- If you'd like to add more tests,
  - Yay!
  - Add to `var tests = […]` (I'm curious what you think of the process of adding more tests)

If you'd like to see what I was working through, there's a spreadsheet called AF Info in the shared Drive directory

## Known issues
- A couple little style concerns (left border of help section doesn't reach the bottom, etc)
- I'm embarrassed to admit that I did not approach this as mobile-first; I still need to look at the smaller layouts
- For testing, frames don't auto advance (think they should?)
- After initial testing, the Next button will become inactive until each frame has an answer
- Close and Start over buttons don't do anything
- Need to think about testing with IE

## Other considerations to test, but probably later
- How does the link behave if there are two on the page?
- Does this addition compete with other modals on the page?
- How does the functionality handle when the class is on something that's not a link?
- @johnnyporkchops suggested being able to add the calculator as a Wagtail module. _mi piace_. Interesting…
- Should we add additional tests?

____
